File: mobile/android/app/src/main/java/com/accord/data/service/WebSocketService.kt
package com.accord.data.service

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * Connection state for the WebSocket.
 */
enum class ConnectionState {
    Disconnected,
    Connecting,
    Connected,
    Reconnecting,
}

/**
 * Sealed hierarchy of messages received from the relay server.
 */
sealed class WsIncoming {
    /** Server confirmed authentication. */
    data class Authenticated(val userId: String) : WsIncoming()

    /** Server hello with version info. */
    data class Hello(
        val serverVersion: String,
        val serverBuildHash: String,
        val protocolVersion: Int,
    ) : WsIncoming()

    /** A channel (or DM) chat message. */
    data class ChannelMessage(
        val from: String,
        val channelId: String,
        val encryptedData: String,
        val messageId: String,
        val timestamp: Long,
        val replyTo: String? = null,
        val senderDisplayName: String? = null,
        val isDm: Boolean = false,
    ) : WsIncoming()

    /** A message was edited. */
    data class MessageEdit(
        val messageId: String,
        val channelId: String,
        val encryptedData: String,
        val editedAt: Long,
    ) : WsIncoming()

    /** A message was deleted. */
    data class MessageDelete(val messageId: String) : WsIncoming()

    /** Presence update for a user. */
    data class PresenceUpdate(val userId: String, val status: String) : WsIncoming()

    /** Bulk presence on connect. */
    data class PresenceBulk(val members: List<Pair<String, String>>) : WsIncoming()

    /** Someone started typing. */
    data class TypingStart(val channelId: String, val userId: String) : WsIncoming()

    /** Reaction added. */
    data class ReactionAdd(
        val messageId: String,
        val reactions: List<Map<String, Any>>,
    ) : WsIncoming()

    /** Reaction removed. */
    data class ReactionRemove(
        val messageId: String,
        val reactions: List<Map<String, Any>>,
    ) : WsIncoming()

    /** Message pinned. */
    data class MessagePin(val messageId: String, val pinnedBy: String, val timestamp: Long) : WsIncoming()

    /** Message unpinned. */
    data class MessageUnpin(val messageId: String) : WsIncoming()

    /** Read receipt. */
    data class ReadReceipt(val channelId: String, val userId: String, val messageId: String, val timestamp: Long) : WsIncoming()

    /** Voice peer joined. */
    data class VoicePeerJoined(val channelId: String, val userId: String) : WsIncoming()

    /** Voice peer left. */
    data class VoicePeerLeft(val channelId: String, val userId: String) : WsIncoming()

    /** Voice signaling (offer/answer/ice). */
    data class VoiceSignaling(val channelId: String, val fromUserId: String, val signalType: String, val data: String) : WsIncoming()

    /** Node created/joined/left confirmations. */
    data class NodeEvent(val type: String, val raw: JSONObject) : WsIncoming()

    /** Channel created. */
    data class ChannelCreated(val raw: JSONObject) : WsIncoming()

    /** Server error. */
    data class Error(val message: String, val code: String? = null) : WsIncoming()

    /** Pong response. */
    data class Pong(val timestamp: Long) : WsIncoming()

    /** Key bundle response. */
    data class KeyBundleResponse(val identityKey: String, val signedPrekey: String, val oneTimePrekey: String?) : WsIncoming()

    /** Prekey messages retrieved. */
    data class PrekeyMessages(val messages: List<JSONObject>) : WsIncoming()

    /** Any unrecognized message — raw JSON. */
    data class Unknown(val raw: JSONObject) : WsIncoming()
}

/**
 * Manages the WebSocket connection to an Accord relay server.
 *
 * Uses post-upgrade authentication (send `{"Authenticate":{"token":"..."}}` within 5s).
 * Auto-reconnects with exponential backoff (1s → 30s max).
 * Exposes connection state as [StateFlow] and incoming messages as [SharedFlow].
 */
class WebSocketService(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob()),
) {
    private val client = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MINUTES) // no read timeout for WS
        .build()

    private var webSocket: WebSocket? = null
    private var authToken: String? = null
    private var relayUrl: String? = null

    // ── Public state ──────────────────────────────────────────────────

    private val _connectionState = MutableStateFlow(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _incoming = MutableSharedFlow<WsIncoming>(
        replay = 0,
        extraBufferCapacity = 256,
    )
    /** Flow of parsed incoming messages from the relay. */
    val incoming: SharedFlow<WsIncoming> = _incoming.asSharedFlow()

    // ── Reconnect state ───────────────────────────────────────────────

    private val reconnectAttempt = AtomicInteger(0)
    private var reconnectJob: Job? = null

    companion object {
        private const val INITIAL_BACKOFF_MS = 1_000L
        private const val MAX_BACKOFF_MS = 30_000L
    }

    // ── Public API ────────────────────────────────────────────────────

    /** Connect to a relay WebSocket endpoint. */
    fun connect(relayUrl: String, authToken: String) {
        disconnect()
        this.authToken = authToken
        this.relayUrl = relayUrl
        reconnectAttempt.set(0)
        doConnect()
    }

    /** Send a typed WS message matching the server's WsMessage envelope. */
    fun sendMessage(messageType: Any): Boolean {
        val envelope = JSONObject().apply {
            put("message_type", when (messageType) {
                is JSONObject -> messageType
                else -> messageType // caller can pass JSONObject directly
            })
            put("message_id", UUID.randomUUID().toString())
            put("timestamp", System.currentTimeMillis() / 1000)
        }
        return send(envelope.toString())
    }

    /** Send a raw JSON string. */
    fun send(json: String): Boolean {
        return webSocket?.send(json) ?: false
    }

    fun disconnect() {
        reconnectJob?.cancel()
        reconnectJob = null
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        _connectionState.value = ConnectionState.Disconnected
    }

    // ── Convenience send helpers ──────────────────────────────────────

    fun sendChannelMessage(channelId: String, encryptedData: String, replyTo: String? = null) {
        val type = JSONObject().apply {
            put("ChannelMessage", JSONObject().apply {
                put("channel_id", channelId)
                put("encrypted_data", encryptedData)
                if (replyTo != null) put("reply_to", replyTo)
            })
        }
        sendEnvelope(type)
    }

    fun sendDirectMessage(toUser: String, encryptedData: String) {
        val type = JSONObject().apply {
            put("DirectMessage", JSONObject().apply {
                put("to_user", toUser)
                put("encrypted_data", encryptedData)
            })
        }
        sendEnvelope(type)
    }

    fun sendTypingStart(channelId: String) {
        val type = JSONObject().apply {
            put("TypingStart", JSONObject().apply {
                put("channel_id", channelId)
            })
        }
        sendEnvelope(type)
    }

    fun sendEditMessage(messageId: String, encryptedData: String) {
        val type = JSONObject().apply {
            put("EditMessage", JSONObject().apply {
                put("message_id", messageId)
                put("encrypted_data", encryptedData)
            })
        }
        sendEnvelope(type)
    }

    fun sendDeleteMessage(messageId: String) {
        val type = JSONObject().apply {
            put("DeleteMessage", JSONObject().apply {
                put("message_id", messageId)
            })
        }
        sendEnvelope(type)
    }

    fun sendAddReaction(messageId: String, emoji: String) {
        val type = JSONObject().apply {
            put("AddReaction", JSONObject().apply {
                put("message_id", messageId)
                put("emoji", emoji)
            })
        }
        sendEnvelope(type)
    }

    fun sendRemoveReaction(messageId: String, emoji: String) {
        val type = JSONObject().apply {
            put("RemoveReaction", JSONObject().apply {
                put("message_id", messageId)
                put("emoji", emoji)
            })
        }
        sendEnvelope(type)
    }

    fun sendPing() {
        val type = JSONObject().apply { put("Ping", JSONObject()) }
        sendEnvelope(type)
    }

    fun joinChannel(channelId: String) {
        val type = JSONObject().apply {
            put("JoinChannel", JSONObject().apply {
                put("channel_id", channelId)
            })
        }
        sendEnvelope(type)
    }

    fun leaveChannel(channelId: String) {
        val type = JSONObject().apply {
            put("LeaveChannel", JSONObject().apply {
                put("channel_id", channelId)
            })
        }
        sendEnvelope(type)
    }

    fun joinVoiceChannel(channelId: String) {
        val type = JSONObject().apply {
            put("JoinVoiceChannel", JSONObject().apply {
                put("channel_id", channelId)
            })
        }
        sendEnvelope(type)
    }

    fun leaveVoiceChannel(channelId: String) {
        val type = JSONObject().apply {
            put("LeaveVoiceChannel", JSONObject().apply {
                put("channel_id", channelId)
            })
        }
        sendEnvelope(type)
    }

    fun publishKeyBundle(identityKey: String, signedPrekey: String, oneTimePrekeys: List<String>) {
        val type = JSONObject().apply {
            put("PublishKeyBundle", JSONObject().apply {
                put("identity_key", identityKey)
                put("signed_prekey", signedPrekey)
                put("one_time_prekeys", org.json.JSONArray(oneTimePrekeys))
            })
        }
        sendEnvelope(type)
    }

    fun fetchKeyBundle(targetUserId: String) {
        val type = JSONObject().apply {
            put("FetchKeyBundle", JSONObject().apply {
                put("target_user_id", targetUserId)
            })
        }
        sendEnvelope(type)
    }

    // ── Internal ──────────────────────────────────────────────────────

    private fun sendEnvelope(messageTypeObj: JSONObject) {
        val envelope = JSONObject().apply {
            put("message_type", messageTypeObj)
            put("message_id", UUID.randomUUID().toString())
            put("timestamp", System.currentTimeMillis() / 1000)
        }
        send(envelope.toString())
    }

    private fun doConnect() {
        val url = relayUrl ?: return
        val wsUrl = url.replace(Regex("^http"), "ws").trimEnd('/') + "/ws"

        _connectionState.value = if (reconnectAttempt.get() > 0) {
            ConnectionState.Reconnecting
        } else {
            ConnectionState.Connecting
        }

        val request = Request.Builder()
            .url(wsUrl)
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                // Post-upgrade auth: send Authenticate message within 5s
                val authMsg = JSONObject().apply {
                    put("Authenticate", JSONObject().apply {
                        put("token", authToken)
                    })
                }
                webSocket.send(authMsg.toString())
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _connectionState.value = ConnectionState.Disconnected
                scope.launch { _incoming.emit(WsIncoming.Error(t.message ?: "Connection failed")) }
                scheduleReconnect()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = ConnectionState.Disconnected
                // Don't reconnect on intentional close or auth failure
                if (code != 1000 && code != 4401 && code != 4403) {
                    scheduleReconnect()
                }
            }
        })
    }

    private fun handleMessage(text: String) {
        val json = try { JSONObject(text) } catch (_: Exception) { return }
        val type = json.optString("type", "")

        val msg: WsIncoming = when (type) {
            "authenticated" -> {
                _connectionState.value = ConnectionState.Connected
                reconnectAttempt.set(0)
                WsIncoming.Authenticated(json.optString("user_id"))
            }

            "hello" -> WsIncoming.Hello(
                serverVersion = json.optString("server_version"),
                serverBuildHash = json.optString("server_build_hash"),
                protocolVersion = json.optInt("protocol_version", 0),
            )

            "channel_message" -> WsIncoming.ChannelMessage(
                from = json.optString("from"),
                channelId = json.optString("channel_id"),
                encryptedData = json.optString("encrypted_data"),
                messageId = json.optString("message_id"),
                timestamp = json.optLong("timestamp"),
                replyTo = json.optString("reply_to").takeIf { it.isNotEmpty() },
                senderDisplayName = json.optString("sender_display_name").takeIf { it.isNotEmpty() },
                isDm = json.optBoolean("is_dm", false),
            )

            "message_edit" -> WsIncoming.MessageEdit(
                messageId = json.optString("message_id"),
                channelId = json.optString("channel_id"),
                encryptedData = json.optString("encrypted_data"),
                editedAt = json.optLong("edited_at"),
            )

            "message_delete" -> WsIncoming.MessageDelete(json.optString("message_id"))

            "presence_update" -> WsIncoming.PresenceUpdate(
                userId = json.optString("user_id"),
                status = json.optString("status"),
            )

            "presence_bulk" -> {
                val arr = json.optJSONArray("members")
                val members = mutableListOf<Pair<String, String>>()
                if (arr != null) {
                    for (i in 0 until arr.length()) {
                        val m = arr.optJSONObject(i) ?: continue
                        members.add(m.optString("user_id") to m.optString("status"))
                    }
                }
                WsIncoming.PresenceBulk(members)
            }

            "typing_start" -> WsIncoming.TypingStart(
                channelId = json.optString("channel_id"),
                userId = json.optString("user_id"),
            )

            "reaction_add" -> WsIncoming.ReactionAdd(
                messageId = json.optString("message_id"),
                reactions = parseReactions(json),
            )

            "reaction_remove" -> WsIncoming.ReactionRemove(
                messageId = json.optString("message_id"),
                reactions = parseReactions(json),
            )

            "message_pin" -> WsIncoming.MessagePin(
                messageId = json.optString("message_id"),
                pinnedBy = json.optString("pinned_by"),
                timestamp = json.optLong("timestamp"),
            )

            "message_unpin" -> WsIncoming.MessageUnpin(json.optString("message_id"))

            "read_receipt" -> WsIncoming.ReadReceipt(
                channelId = json.optString("channel_id"),
                userId = json.optString("user_id"),
                messageId = json.optString("message_id"),
                timestamp = json.optLong("timestamp"),
            )

            "voice_peer_joined" -> WsIncoming.VoicePeerJoined(
                channelId = json.optString("channel_id"),
                userId = json.optString("user_id"),
            )

            "voice_peer_left" -> WsIncoming.VoicePeerLeft(
                channelId = json.optString("channel_id"),
                userId = json.optString("user_id"),
            )

            "voice_offer", "voice_answer", "voice_ice_candidate", "p2p_signal" -> WsIncoming.VoiceSignaling(
                channelId = json.optString("channel_id"),
                fromUserId = json.optString("from_user_id", json.optString("user_id")),
                signalType = type,
                data = text,
            )

            "node_created", "node_joined", "node_left", "node_info" -> WsIncoming.NodeEvent(type, json)

            "channel_created" -> WsIncoming.ChannelCreated(json)

            "key_bundle_response" -> WsIncoming.KeyBundleResponse(
                identityKey = json.optString("identity_key"),
                signedPrekey = json.optString("signed_prekey"),
                oneTimePrekey = json.optString("one_time_prekey").takeIf { it.isNotEmpty() },
            )

            "prekey_messages" -> {
                val arr = json.optJSONArray("messages")
                val msgs = mutableListOf<JSONObject>()
                if (arr != null) {
                    for (i in 0 until arr.length()) {
                        arr.optJSONObject(i)?.let { msgs.add(it) }
                    }
                }
                WsIncoming.PrekeyMessages(msgs)
            }

            "pong" -> WsIncoming.Pong(json.optLong("timestamp"))

            "error" -> WsIncoming.Error(
                message = json.optString("message", json.optString("error", "Unknown error")),
                code = json.optString("code").takeIf { it.isNotEmpty() },
            )

            else -> WsIncoming.Unknown(json)
        }

        scope.launch { _incoming.emit(msg) }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseReactions(json: JSONObject): List<Map<String, Any>> {
        val arr = json.optJSONArray("reactions") ?: return emptyList()
        val result = mutableListOf<Map<String, Any>>()
        for (i in 0 until arr.length()) {
            val obj = arr.optJSONObject(i) ?: continue
            val map = mutableMapOf<String, Any>()
            for (key in obj.keys()) {
                map[key] = obj.get(key)
            }
            result.add(map)
        }
        return result
    }

    private fun scheduleReconnect() {
        reconnectJob?.cancel()
        val attempt = reconnectAttempt.incrementAndGet()
        val delay = (INITIAL_BACKOFF_MS * (1L shl (attempt - 1).coerceAtMost(5)))
            .coerceAtMost(MAX_BACKOFF_MS)

        reconnectJob = scope.launch {
            delay(delay)
            if (isActive) {
                _connectionState.value = ConnectionState.Reconnecting
                doConnect()
            }
        }
    }
}

File: mobile/ios/Accord/Services/WebSocketService.swift
// WebSocketService.swift — WebSocket connection to the Accord relay

import Foundation
import Combine

// MARK: - Connection State

enum ConnectionState: String {
    case disconnected
    case connecting
    case connected
    case reconnecting
}

// MARK: - Incoming WS Message Types

/// Parsed server→client WebSocket messages
enum WSIncomingMessage {
    case authenticated(userId: String)
    case channelMessage(ChannelMessagePayload)
    case presenceUpdate(userId: String, status: String)
    case presenceBulk(members: [[String: Any]])
    case typingStart(channelId: String, userId: String, publicKeyHash: String?)
    case messageEdit(messageId: String, channelId: String, encryptedData: String, editedAt: Int)
    case messageDelete(messageId: String, channelId: String)
    case reactionAdd(messageId: String, reactions: [[String: Any]])
    case reactionRemove(messageId: String, reactions: [[String: Any]])
    case messagePin(messageId: String, timestamp: Int, pinnedBy: String)
    case messageUnpin(messageId: String)
    case readReceipt(channelId: String, userId: String, messageId: String, timestamp: Int)
    case hello(serverVersion: String?, serverBuildHash: String?)
    case error(code: String?, message: String)
    case unknown(type: String, raw: [String: Any])
}

struct ChannelMessagePayload {
    let messageId: String
    let channelId: String
    let from: String?
    let encryptedData: String
    let timestamp: Int
    let isDM: Bool
    let replyTo: String?
    let repliedMessage: [String: Any]?
    let senderDisplayName: String?
}

// MARK: - WebSocketService

/// Handles persistent WebSocket connection to the relay for real-time events.
@Observable
final class WebSocketService {
    private let relayURL: String
    private var authToken: String?
    private var webSocketTask: URLSessionWebSocketTask?
    private let session: URLSession

    // Connection state
    private(set) var connectionState: ConnectionState = .disconnected
    var isConnected: Bool { connectionState == .connected }

    // Reconnection
    private var reconnectAttempt: Int = 0
    private let maxReconnectAttempts: Int = 20
    private var reconnectDelay: TimeInterval = 1.0
    private let maxReconnectDelay: TimeInterval = 30.0
    private var reconnectTask: Task<Void, Never>?
    private var isDestroyed: Bool = false

    // Ping keepalive
    private var pingTask: Task<Void, Never>?

    // MARK: - Callbacks

    var onMessage: ((WSIncomingMessage) -> Void)?
    var onRawText: ((String) -> Void)?
    var onConnectionStateChanged: ((ConnectionState) -> Void)?

    // Combine publishers (alternative to callbacks)
    let messagePublisher = PassthroughSubject<WSIncomingMessage, Never>()
    let connectionStatePublisher = PassthroughSubject<ConnectionState, Never>()

    init(relayURL: String) {
        self.relayURL = relayURL
        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = true
        self.session = URLSession(configuration: config)
    }

    // MARK: - Connection

    func connect(token: String) {
        guard !isDestroyed else { return }
        self.authToken = token

        // Build WS URL: replace http(s) with ws(s), append /ws
        var wsURLString = relayURL
        if wsURLString.hasPrefix("https") {
            wsURLString = "wss" + wsURLString.dropFirst(5)
        } else if wsURLString.hasPrefix("http") {
            wsURLString = "ws" + wsURLString.dropFirst(4)
        }
        if !wsURLString.hasSuffix("/ws") {
            wsURLString += "/ws"
        }

        guard let url = URL(string: wsURLString) else {
            print("[WS] Invalid URL: \(wsURLString)")
            return
        }

        setConnectionState(.connecting)

        let task = session.webSocketTask(with: url)
        self.webSocketTask = task
        task.resume()

        // Post-upgrade auth: send Authenticate message immediately
        sendAuthMessage(token: token)
    }

    func disconnect() {
        isDestroyed = true
        reconnectTask?.cancel()
        reconnectTask = nil
        pingTask?.cancel()
        pingTask = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        setConnectionState(.disconnected)
    }

    func reconnectWithToken(_ token: String) {
        isDestroyed = false
        reconnectAttempt = 0
        reconnectDelay = 1.0
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        connect(token: token)
    }

    // MARK: - Send

    /// Send a typed WS message as JSON
    func send(_ messageType: [String: Any]) async throws {
        let envelope: [String: Any] = [
            "message_type": messageType,
            "message_id": UUID().uuidString,
            "timestamp": Int(Date().timeIntervalSince1970)
        ]
        let data = try JSONSerialization.data(withJSONObject: envelope)
        guard let text = String(data: data, encoding: .utf8) else { return }
        try await sendRaw(text)
    }

    /// Send a channel message
    func sendChannelMessage(channelId: String, encryptedData: String, replyTo: String? = nil) async throws {
        var msg: [String: Any] = [
            "ChannelMessage": [
                "channel_id": channelId,
                "encrypted_data": encryptedData,
            ] as [String: Any]
        ]
        if let replyTo {
            var inner = msg["ChannelMessage"] as! [String: Any]
            inner["reply_to"] = replyTo
            msg["ChannelMessage"] = inner
        }
        try await send(msg)
    }

    /// Send typing start indicator
    func sendTypingStart(channelId: String) async throws {
        try await send(["TypingStart": ["channel_id": channelId]])
    }

    /// Send ping
    func sendPing() async throws {
        try await send(["Ping": [:] as [String: Any]])
    }

    /// Send raw text
    func sendRaw(_ text: String) async throws {
        guard let task = webSocketTask else {
            throw URLError(.badServerResponse)
        }
        try await task.send(.string(text))
    }

    // MARK: - Private

    private func sendAuthMessage(token: String) {
        let authJSON = "{\"Authenticate\":{\"token\":\"\(token)\"}}"
        webSocketTask?.send(.string(authJSON)) { [weak self] error in
            if let error {
                print("[WS] Failed to send auth message: \(error)")
                self?.handleDisconnect()
            } else {
                // Start listening — server will send "authenticated" or "error"
                self?.listenForMessages()
            }
        }
    }

    private func listenForMessages() {
        webSocketTask?.receive { [weak self] result in
            guard let self, !self.isDestroyed else { return }
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self.handleTextMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self.handleTextMessage(text)
                    }
                @unknown default:
                    break
                }
                self.listenForMessages()
            case .failure(let error):
                print("[WS] Receive error: \(error)")
                self.handleDisconnect()
            }
        }
    }

    private func handleTextMessage(_ text: String) {
        onRawText?(text)

        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            return
        }

        let parsed: WSIncomingMessage

        switch type {
        case "authenticated":
            let userId = json["user_id"] as? String ?? ""
            parsed = .authenticated(userId: userId)
            // Auth confirmed — we're truly connected
            reconnectAttempt = 0
            reconnectDelay = 1.0
            setConnectionState(.connected)
            startPingKeepalive()

        case "channel_message":
            parsed = .channelMessage(ChannelMessagePayload(
                messageId: json["message_id"] as? String ?? UUID().uuidString,
                channelId: json["channel_id"] as? String ?? "",
                from: json["from"] as? String,
                encryptedData: json["encrypted_data"] as? String ?? "",
                timestamp: json["timestamp"] as? Int ?? Int(Date().timeIntervalSince1970),
                isDM: json["is_dm"] as? Bool ?? false,
                replyTo: json["reply_to"] as? String,
                repliedMessage: json["replied_message"] as? [String: Any],
                senderDisplayName: json["sender_display_name"] as? String
            ))

        case "presence_update":
            parsed = .presenceUpdate(
                userId: json["user_id"] as? String ?? "",
                status: json["status"] as? String ?? "offline"
            )

        case "presence_bulk":
            parsed = .presenceBulk(members: json["members"] as? [[String: Any]] ?? [])

        case "typing_start":
            parsed = .typingStart(
                channelId: json["channel_id"] as? String ?? "",
                userId: json["user_id"] as? String ?? "",
                publicKeyHash: json["public_key_hash"] as? String
            )

        case "message_edit":
            parsed = .messageEdit(
                messageId: json["message_id"] as? String ?? "",
                channelId: json["channel_id"] as? String ?? "",
                encryptedData: json["encrypted_data"] as? String ?? "",
                editedAt: json["edited_at"] as? Int ?? 0
            )

        case "message_delete":
            parsed = .messageDelete(
                messageId: json["message_id"] as? String ?? "",
                channelId: json["channel_id"] as? String ?? ""
            )

        case "reaction_add":
            parsed = .reactionAdd(
                messageId: json["message_id"] as? String ?? "",
                reactions: json["reactions"] as? [[String: Any]] ?? []
            )

        case "reaction_remove":
            parsed = .reactionRemove(
                messageId: json["message_id"] as? String ?? "",
                reactions: json["reactions"] as? [[String: Any]] ?? []
            )

        case "message_pin":
            parsed = .messagePin(
                messageId: json["message_id"] as? String ?? "",
                timestamp: json["timestamp"] as? Int ?? 0,
                pinnedBy: json["pinned_by"] as? String ?? ""
            )

        case "message_unpin":
            parsed = .messageUnpin(messageId: json["message_id"] as? String ?? "")

        case "read_receipt":
            parsed = .readReceipt(
                channelId: json["channel_id"] as? String ?? "",
                userId: json["user_id"] as? String ?? "",
                messageId: json["message_id"] as? String ?? "",
                timestamp: json["timestamp"] as? Int ?? 0
            )

        case "hello":
            parsed = .hello(
                serverVersion: json["server_version"] as? String,
                serverBuildHash: json["server_build_hash"] as? String
            )

        case "error":
            let code = json["code"] as? String
            let message = json["message"] as? String ?? json["error"] as? String ?? "Unknown error"
            parsed = .error(code: code, message: message)

            // Handle auth errors
            if code == "auth_failed" {
                handleDisconnect()
                return
            }

        default:
            parsed = .unknown(type: type, raw: json)
        }

        onMessage?(parsed)
        messagePublisher.send(parsed)
    }

    private func handleDisconnect() {
        pingTask?.cancel()
        pingTask = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil

        guard !isDestroyed else {
            setConnectionState(.disconnected)
            return
        }

        guard reconnectAttempt < maxReconnectAttempts else {
            setConnectionState(.disconnected)
            print("[WS] Max reconnect attempts reached")
            return
        }

        setConnectionState(.reconnecting)
        scheduleReconnect()
    }

    private func scheduleReconnect() {
        let delay = reconnectDelay
        reconnectAttempt += 1
        // Exponential backoff: 1, 2, 4, 8, ..., max 30s
        reconnectDelay = min(reconnectDelay * 2, maxReconnectDelay)

        print("[WS] Reconnecting in \(delay)s (attempt \(reconnectAttempt)/\(maxReconnectAttempts))")

        reconnectTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard let self, !self.isDestroyed, !Task.isCancelled else { return }
            guard let token = self.authToken else { return }
            self.connect(token: token)
        }
    }

    private func startPingKeepalive() {
        pingTask?.cancel()
        pingTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 30_000_000_000) // 30s
                guard let self, self.isConnected else { break }
                try? await self.sendPing()
            }
        }
    }

    private func setConnectionState(_ state: ConnectionState) {
        guard connectionState != state else { return }
        connectionState = state
        onConnectionStateChanged?(state)
        connectionStatePublisher.send(state)
    }
}
